### ISSUE TITLE
generate a git style diff for strings

### ISSUE THREAD

**Issue by nhooyr** [google/go-cmp#103]
It'd be nice to have git style diffs between multilined strings.

**Comment by dvyukov** [comment]
If there would be an option that would make multi-line strings output as actually multi-line, rather then escaped with `\n`, that would solve part of the problem for me. Currently I have not too useful:
```
		bisect_test.go:264: {string}:
			-: "syzbot has bisected this bug to:\n\ncommit 111111111111111111111111\nAuthor: Hacker Kernelov <hacker@kernel.org>\nDate:   Wed Feb 9 04:05:06 2000 +0000\n\n    kernel: break build\n\nbisection log:  https://testapp.appspot.com/x/bisect.txt?x=17000000000000\nstart commit:   11111111 kernel_commit_title1\ngit tree:       repo1 branch1\nfinal crash:    https://testapp.appspot.com/x/report.txt?x=14800000000000\nconsole output: https://testapp.appspot.com/x/log.txt?x=10800000000000\nkernel config:  https://testapp.appspot.com/x/.config?x=4d11162a90e18f28\ndashboard link: https://testapp.appspot.com/bug?extid=d21188062e41595c9273\nsyz repro:      https://testapp.appspot.com/x/repro.syz?x=16000000000000\nC reproducer:   https://testapp.appspot.com/x/repro.c?x=11000000000000\n\nReported-by: syzbot+d21188062e41595c9273@testapp.appspotmail.com\nFixes: 11111111 (\"kernel: break build\")\n"
			+: "syzbot has bisected this bug to:\n\ncommit 36e65cb4a0448942ec316b24d60446bbd5cc7827\nAuthor: Author Kernelov <author@kernel.org>\nDate:   Wed Feb 9 04:05:06 2000 +0000\n\n    kernel: add a bug\n\nbisection log:  https://testapp.appspot.com/x/bisect.txt?x=17000000000000\nstart commit:   11111111 kernel_commit_title1\ngit tree:       repo1 branch1\nfinal crash:    https://testapp.appspot.com/x/report.txt?x=14800000000000\nconsole output: https://testapp.appspot.com/x/log.txt?x=10800000000000\nkernel config:  https://testapp.appspot.com/x/.config?x=4d11162a90e18f28\ndashboard link: https://testapp.appspot.com/bug?extid=d21188062e41595c9273\nsyz repro:      https://testapp.appspot.com/x/repro.syz?x=16000000000000\nC reproducer:   https://testapp.appspot.com/x/repro.c?x=11000000000000\n\nReported-by: syzbot+d21188062e41595c9273@testapp.appspotmail.com\nFixes: 36e65cb4 (\"kernel: add a bug\")\n"
```

**Comment by dnephin** [comment]
While you are waiting on a response you might be interested in https://godoc.org/gotest.tools/assert#Equal  (unified diffs are git-style diffs)
> If either x or y are a multi-line string the failure message will include a unified diff of the two values. If the values only differ by whitespace the unified diff will be augmented by replacing whitespace characters with visible characters to identify the whitespace difference.

In the same package there is https://godoc.org/gotest.tools/assert#DeepEqual which uses `google/go-cmp`.

**Comment by nhooyr** [comment]
awesome, very happy to see this closed.

Nice to see how the implementation goes one step further. A multiline string really is just a slice of strings so it makes sense to generalize the diff to slices.

**Comment by dvyukov** [comment]
Cool! Thanks!

```
		  	... // 15 identical lines
		  	"bisection log:  https://testapp.appspot.com/x/bisect.txt?x=17000000000000",
		  	"final crash:    https://testapp.appspot.com/x/report.txt?x=14800000000000",
		- 	"console output: https://testapp.appspot.com/x/log.txt?x=10800000000000",
		+ 	"console output1: https://testapp.appspot.com/x/log.txt?x=10800000000000",
		  	"",
		  	"IMPORTANT: if you fix the bug, please add the following tag to the commit:",
		  	... // 13 identical lines
```
